### DISCUSSION TITLE
How to use edgex with spiffe service on multiple nodes?

### DISCUSSION THREAD

**Discussion by LavenderQAQ** [edgexfoundry/community#45] [Q&A] [answered]
Hello, I'm from the `openyurt` project, I'm trying to deploy a secure version of edgex on k8s, but I'm having some problems.

Edgex's service by default requires a shared volume to distribute tokens (almost every service has /tmp/edgex/secrets mounted), and edgex-init is required to generate scripts when security versions are started. In the environment is not enabled PVC k8s unable to use a shared file system to complete the operation, but we found at [spiffe](https://docs.edgexfoundry.org/2.2/design/adr/security/0020-spiffe/).

Could you please tell me how I can test it using `edgex-compose` and whether it can get rid of the constraints of the shared file system?

**Comment by lenny-goodell** [comment, marked as answer]
Please see the Helm example here for reference on K8s deployment.
https://github.com/edgexfoundry/edgex-examples/tree/main/deployment/helm

**Reply by LavenderQAQ** [reply]
Thank you. I've seen this repo before. I would like to ask if there is any method that does not use pvc at present. Because our project does not want to use pv and pvc at present, I would like to ask. Can the spiffe provider be used to distribute tokens currently？

**Reply by bnevis-i** [reply]
@LavenderQAQ 

I haven't gotten around to integrating SPIFFE/SPIRE support into the helm chart.

Feel free to jump in: https://github.com/edgexfoundry/edgex-examples/issues/94

There are two use cases for shared volumes:

(1) Distributing the secret store tokens to the EdgeX microservices

(2) Distributing custom entrypoint scripts when running in security-enabled mode


Thus, even if you solved (1) with SPIFFE/SPIRE support, you would still have to deal with (2).

(2) could be dealt with by putting the entrypoint scripts into configmaps and mounting those into the containers, and have kubernetes distribute them.  But it would increase the maintenance burden by having docker and k8s do it in different ways.

One could avoid (1) and (2) by running with no security and using a Kubernetes service mesh or other networking policies to firewall off EdgeX from other parts of the k8s cluster  (basically, implement security yourself).

None of these problems are insurmountable with a sufficient amount of time and effort.

**Reply by bnevis-i** [reply]
@LavenderQAQ

Thank you very much for bringing it up and for supporting EdgeX in OpenYurt.  I look forward to building a better integration.

**Reply by LavenderQAQ** [reply]
@bnevis-i It's my pleasure. Thank you very much for your detailed answer. Indeed, script distribution and token distribution without pvc will bring great maintenance burden. Our project will also discuss this problem and choose a better solution. Thank you again for your answer.

**Reply by bnevis-i** [reply]
> @bnevis-i It's my pleasure. Thank you very much for your detailed answer. Indeed, script distribution and token distribution without pvc will bring great maintenance burden. Our project will also discuss this problem and choose a better solution. Thank you again for your answer.

Please share your ideas for improvement.  If you have a good idea it will reduce your maintenance burden to push the solutions upstream.

**Reply by bnevis-i** [reply]
PS:
- [helm deployment: Eliminate pre-shared secrets RWX volumes (require only RWO volumes)](https://github.com/edgexfoundry/edgex-examples/issues/95)
- [helm deployment: Eliminate shared wrapper scripts RWX volume](https://github.com/edgexfoundry/edgex-examples/issues/100)
- [helm deployment: Support Istio service mesh](https://github.com/edgexfoundry/edgex-examples/issues/99)
- [helm deployment: Add network policy declarations to PodSpecs](https://github.com/edgexfoundry/edgex-examples/issues/98)

**Reply by LavenderQAQ** [reply]
Ok, we plan to try to adapt the secure version after we complete the proposal of synchronizing edgex with openyurt. Thank you again for your help.

**Reply by bnevis-i** [reply]
I also see a possible future where we use Kubernetes service account authentication to talk to a cluster-based Vault instance.  In this scenario, we could have an init container that pre-seeds the token, and getting rid of the pre-shared secrets volume that way.  There's also an issue filed for [that](https://github.com/edgexfoundry/edgex-examples/issues/96)

**Reply by bnevis-i** [reply]
@LavenderQAQ   A simpler stopgap method to achieve what you want might be to write a file system watcher, that when an init script for Vault token is written to disk, slurps it up, sticks it in a configmap or secret via the kubernetes API, and then the remote target container consumes it via a secretRef or configmap Ref, where Kubernetes is doing the hard work of distribution.

**Reply by LavenderQAQ** [reply]
@bnevis-i Yeah, yeah, it looks like we need the init container for script distribution. This would be much easier if you had a cluster-level vault(for token distribution). For now, I'd like to use an init container for script distribution. Of course I need to discuss this with the openyurt community first, and if this goes well I would be willing to help edgex-helm repo.

**Reply by LavenderQAQ** [reply]
@bnevis-i We did consider writing a file system watcher before😆, but we feel this one really didn't look that beautiful. We're still trying not to do this as much as possible.

**Reply by LavenderQAQ** [reply]
@bnevis-i Personally, I currently feel that the init container distributing scripts and using spiffe to distribute tokens seems like a good solution. It seems a little more elegant🧐.
